drafts/maui-test_pt1.md

```diff
@@ -0,0 +1,332 @@
+
+Consider the following class:
+
+``` C#
+namespace CatFinder.Services.FileSystemServices;
+
+public sealed class FileSystemCsvService : ICsvService
+{
+
+    private const string CountriesCsvFileName = "countries.csv";
+    private readonly IFileSystem _fileSystem;
+
+    public FileSystemCsvService(IFileSystem fileSystem)
+    {
+        _fileSystem = fileSystem;
+    }
+
+    public async Task<IList<Country>> GetCountriesAsync()
+    {
+        try
+        {
+            const char csvSplitChar = ',';
+            const int expectedCsvRowsCount = 4;
+            using var stream = await _fileSystem.OpenAppPackageFileAsync(CountriesCsvFileName);
+            using var reader = new StreamReader(stream);
+            var line = reader.ReadLine();
+            var results = new List<Country>();
+            while (!reader.EndOfStream)
+            {
+                line = reader.ReadLine();
+                var values = line.Split(csvSplitChar);
+                if(values.Length != 4) throw new ArgumentException($"{CountriesCsvFileName} needs to have {expectedCsvRowsCount} rows.");
+                results.Add(new Country(values[0], double.Parse(values[1]), double.Parse(values[2]), values[3]));
+            }
+            return results;
+        }
+        catch (Exception ex)
+        {
+            throw new FileSystemCsvServiceException(ex.Message, ex);
+
+        }
+
+    }
+}
+```
+
+The class above shows a class with one method. This method reads a CSV file which contains 244 countries and maps them to Country objects. The CSV file entries look like this:
+
+```
+country,latitude,longitude,name
+AD,42.546245,1.601554,Andorra
+AE,23.424076,53.847818,United Arab Emirates
+AF,33.93911,67.709953,Afghanistan
+...
+```
+
+Now, let us test whether the method loads the correct number of countries and throws the correct exception.
+
+``` C#
+namespace CatFinder.Tests.Services;
+
+public sealed class FileSystemCsvServiceTests
+{
+
+    [Fact]
+    public async void GetCountriesAsync()
+    {
+        var fileSystemMock = GetFileSystemMock();
+        var fileSystemCsvService = new FileSystemCsvService(fileSystemMock.Object);
+
+        var countries = await fileSystemCsvService.GetCountriesAsync();
+
+        const int expectedCountriesCount = 244;
+        countries.Should().NotBeNull();
+        countries.Should().HaveCount(expectedCountriesCount);
+        fileSystemMock.Verify(fileSystem => fileSystem.OpenAppPackageFileAsync(It.IsAny<string>()), Times.Once());
+    }
+}
+```
+
+The first test above checks whether the method loads the correct number of countries. For this, a filesystem mock is needed. You can not access a file via the filesystem of .NET MAUI this will lead to an exception. The reason is that in the test environment, the platform of the filesystem is not determined. To circumvent this, I created a mock object and extracted it to a static class called MockHelper.
```

Comment:
Honest question, not a claim: Would it help readers to understand that this is because the test is run as a unit test without using the platform as in UI tests?

Reply:
Yes, it would. I changed it, thanks.